File: 프로그래머스/2/42587. 프로세스/프로세스.py
from collections import deque

def solution(priorities, location):
    answer = 0
    result = []
    process_list = [i for i in range(len(priorities))]
    
    while priorities:
        now_pro = priorities.pop(0)
        if priorities:  # priorities가 비어 있지 않을 때만 max를 호출
            if now_pro >= max(priorities):
                x = process_list.pop(0)
                result.append(x)
                if x == location:
                    return len(result)
            else:
                priorities.append(now_pro)
                x = process_list.pop(0)
                process_list.append(x)
        else:
            # 이 부분은 priorities가 비어있는 경우입니다.
            # 만약 priorities가 비어 있다면 now_pro는 마지막 원소가 되어야 하므로 
            # max와 같은 비교를 하지 않고 단순히 마지막 원소를 처리합니다.
            x = process_list.pop(0)
            result.append(x)
            if x == location:
                return len(result)

    return result

File: 프로그래머스/2/42587. 프로세스/프로세스.java
import java.util.*;

class Solution {
    
    class Process {
        int priority;
        int idx;
        
        Process(int priority, int idx) {
            this.priority = priority;
            this.idx = idx;
        }
    }
    
    TreeSet<Integer> maxSet = new TreeSet<>(Collections.reverseOrder());
    Map<Integer, Integer> map = new HashMap<>();
    
    public int solution(int[] priorities, int location) {
        Deque<Process> q = new ArrayDeque<>();
        
        for (int i = 0; i < priorities.length; i++) {
            int num = priorities[i];
            q.offer(new Process(num, i));
            maxSet.add(num);
            map.put(num, map.getOrDefault(num, 0) + 1);
        }
        
        int curMax = maxSet.pollFirst();
        int cnt = 0;
        while (!q.isEmpty()) {
            Process cur = q.poll();
            int priority = cur.priority;
            int idx = cur.idx;
            
            if (priority == curMax) {
                if (idx == location) return cnt + 1;
                map.put(priority, map.get(priority) - 1);
                cnt++;
                if (map.get(priority) == 0) curMax = maxSet.pollFirst();
            } else {
                q.offer(cur);
            }
        }
        
        return priorities.length;
    }
}